lib/streamlit/state/session_state.py

```diff
@@ -297,45 +307,109 @@ def _safe_widget_state(self) -> Dict[str, Any]:
 
     @property
     def _merged_state(self) -> Dict[str, Any]:
-        # NOTE: The order that the dicts are unpacked here is important as it
-        #       is what ensures that the new values take priority
-        return {
-            **self._old_state,
-            **self._safe_widget_state(),
-            **self._new_session_state,
-        }
+        return {k: self[k] for k in self}
 
     @property
     def filtered_state(self) -> Dict[str, Any]:
-        return {
-            k: v
-            for k, v in self._merged_state.items()
-            if (
-                not k.startswith(GENERATED_WIDGET_KEY_PREFIX)
-                and not k.startswith(STREAMLIT_INTERNAL_KEY_PREFIX)
-            )
+        """The combined session and widget state, excluding keyless widgets."""
+
+        wid_key_map = self.reverse_key_wid_map
+
+        state: Dict[str, Any] = {}
+        for k, v in self.items():
+            if not is_widget_id(k) and not is_internal_key(k):
+                state[k] = v
+            elif is_keyed_widget_id(k):
+                try:
+                    key = wid_key_map[k]
+                    state[key] = v
+                except KeyError:
+                    # Widget id no longer maps to a key, it is a not yet
+                    # cleared value in old state for a reset widget
+                    pass
+
+        return state
+
+    @property
+    def reverse_key_wid_map(self) -> Dict[str, str]:
+        wid_key_map = {v: k for k, v in self._key_id_mapping.items()}
+        return wid_key_map
+
+    def keys(self) -> Set[str]:
+        """All keys active in Session State, with widget keys converted
+        to widget ids when one is known."""
+        old_keys = set(self._old_state.keys())
+        new_widget_keys = set(self._new_widget_state.keys())
+        new_session_state_keys = {
+            self._get_widget_id(k) for k in self._new_session_state.keys()
         }
+        return old_keys | new_widget_keys | new_session_state_keys
 
     def is_new_state_value(self, key: str) -> bool:
         return key in self._new_session_state
 
+    def is_new_widget_value(self, key: str) -> bool:
+        return key in self._new_widget_state
+
     def __iter__(self) -> Iterator[Any]:
-        return iter(self._merged_state)
+        return iter(self.keys())
 
     def __len__(self) -> int:
-        return len(self._merged_state)
+        return len(self.keys())
 
     def __str__(self):
         return str(self._merged_state)
 
     def __getitem__(self, key: str) -> Any:
+        wid_key_map = self.reverse_key_wid_map
+        widget_id = self._get_widget_id(key)
+        if widget_id in wid_key_map and widget_id == key:
```

Comment:
It took me a minute to understand what the significance of the `widget_id == key` clause is here, so it may be helpful to add a comment explaining that if this is the case, we were given a `widget_id` to begin with